### PR TITLE
Generate Netplan bond interfaces from NetBox LAGs

### LINKED ISSUE #1369: Support creating active/passive bond interfaces from netbox
Sample target netplan config:
```
network:
  version: 2
  ethernets:
    eno8303:
      dhcp4: false
    eno8403:
      dhcp4: false
  bonds:
    bond0:
      interfaces:
        - eno8303
        - eno8403
      parameters:
        mode: active-backup
        primary: eno8303
        mii-monitor-interval: 100
        fail-over-mac-policy: active
```

### PR DESCRIPTION
**Pull request by berendt**
## Summary

Auto-generate a `network_bonds` section in the Netplan parameters when a device has a **port channel / LAG configured in NetBox**. This mirrors the NetBox modelling that SONiC port channel detection relies on (`osism/python-osism`, `detect_port_channels()`): a LAG is an interface of `type == "lag"` and member interfaces carry a `lag` back-reference to it. Only the **host side** is handled here — the SONiC counterpart is unaffected.

## Behaviour

- **LAG interface** (`type=lag`, `managed-by-osism` tag) → `network_bonds` entry. The bond carries the IP addresses and MTU of the LAG interface itself.
- **Member interfaces** (their `lag` field points at a managed LAG) are still rendered as `network_ethernets` entries (MAC `match` + `set-name`), but **without** addresses, DHCP or leaf link-local settings — the bond holds the IP configuration.
- **Default = LACP (802.3ad)**: `lacp-rate: fast`, `mii-monitor-interval: 100`, `transmit-hash-policy: layer3+4`.
- **Per-LAG override** via the `netplan_parameters` custom field on the LAG interface (shallow update, consistent with other interface types). Providing a `parameters` dict replaces the auto-generated defaults entirely — e.g. the `active-backup` case from the issue.
- VRF assignment of the LAG interface is reflected in `network_vrfs`.

The output key `network_bonds` is already rendered to `network.bonds:` by the consuming role template in `osism/ansible-collection-commons` (`roles/network/templates/netplan/01-osism.yaml.j2`), so no change on the consumer side is needed.

### Example output (auto-generated, LACP default)

```yaml
network_ethernets:
  eno8303: {match: {macaddress: "aa:bb:cc:00:00:01"}, set-name: eno8303, mtu: 9100}
  eno8403: {match: {macaddress: "aa:bb:cc:00:00:02"}, set-name: eno8403, mtu: 9100}
network_bonds:
  bond0:
    interfaces: [eno8303, eno8403]
    parameters: {mode: 802.3ad, lacp-rate: fast, mii-monitor-interval: 100, transmit-hash-policy: layer3+4}
    mtu: 9100
    addresses: [10.0.0.5/24]
```

### active-backup case (issue example) — set on the LAG interface's `netplan_parameters` custom field

```json
{"parameters": {"mode": "active-backup", "primary": "eno8303", "mii-monitor-interval": 100, "fail-over-mac-policy": "active"}}
```

## Tests

New `tests/unit/netbox/test_netplan_extractor.py` covering the LACP default, the active-backup custom-field override, and the fallback when the LAG parent is not managed. Full unit suite passes (176 tests).

Closes osism/issues#1369

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: files/netbox/extractors/netplan_extractor.py [hunk 1]
```diff
@@ -155,6 +163,25 @@ def extract(
...
+        # detection relies on.
+        lag_interfaces_by_id = {}
+        for interface in interfaces:
+            if (
+                interface.type
+                and interface.type.value == "lag"
+                and getattr(interface, "tags", None)
+                and "managed-by-osism" in [tag.slug for tag in interface.tags]
+            ):
+                lag_interfaces_by_id[interface.id] = interface
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py]: **Finding 1 (Medium-High) — LAG pre-pass bypasses `enabled` / `mgmt_only` gates.**

This pre-pass only filters on `type == "lag"` + tag, but the post-loop bond assembly at `:730` emits every entry here unconditionally. It skips the `mgmt_only` gate (`:195`) and the disabled-interface gate (`:428`) that all other interfaces pass through. Net effects I confirmed: (1) a `mgmt_only` LAG is emitted as an active bond, and because its VRF assignment is recorded only *after* the `mgmt_only continue` at `:195`, the bond is emitted *without* its VRF membership — inconsistent partial state; (2) a disabled LAG is written both as an `activation-mode: off` ethernet and as an active bond with the same name (collision). Please apply the same `enabled`/`mgmt_only`/name eligibility predicate here that the main loop uses, and add tests for a `mgmt_only` LAG and a disabled LAG.

File: files/netbox/extractors/netplan_extractor.py [hunk 2]
```diff
@@ -405,6 +437,25 @@ def extract(
             else:
                 interface_config["mtu"] = effective_default_mtu
 
+            # Bond / port channel members only need the MAC-based rename and the
+            # MTU here. The bond interface itself carries the IP configuration, so
+            # members must not get addresses, DHCP or leaf link-local settings.
+            lag_parent_id = (
+                interface.lag.id if getattr(interface, "lag", None) else None
+            )
+            if lag_parent_id is not None and lag_parent_id in lag_interfaces_by_id:
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py]: **Finding 2 (Medium) — Disabled member becomes an orphan ethernet, dropped from the bond.**

The disabled-interface branch above (`:427-432`) `continue`s before this LAG-member block, so a disabled member of a managed LAG is written there as a standalone `activation-mode: off` ethernet and never added to `bond_members` — it silently vanishes from the bond's `interfaces` list and becomes an orphan. This is a separate fix from the pre-pass gating: make the disabled check LAG-member-aware (or move the membership detection above it) so a disabled member is excluded from the bond's `interfaces` (and logged) rather than emitted as a dangling ethernet. Needs a test with one disabled member of a two-member bond.

File: files/netbox/extractors/netplan_extractor.py [hunk 3]
```diff
@@ -405,6 +437,25 @@ def extract(
...
+                if hasattr(interface, "custom_fields") and interface.custom_fields:
+                    member_params = interface.custom_fields.get("netplan_parameters")
+                    if member_params and isinstance(member_params, dict):
+                        interface_config.update(member_params)
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py]: **Finding 4 (Medium) — Member `netplan_parameters` shallow-merge can reintroduce addresses/DHCP and clobber `match`/`set-name`.**

This raw `update(member_params)` merges into a dict that already holds `match`/`set-name` and is documented to carry *no* IP/DHCP/link-local config. A member's `netplan_parameters` custom field can therefore (a) reintroduce `addresses`/`dhcp4`/`dhcp6`/`link-local` — violating the "members carry no addresses" invariant stated in the docstring and CLAUDE.md — and (b) overwrite `match`/`set-name`, breaking the MAC-rename identity. Either filter the conflicting keys here for the member path, or document this as an intentional override. A test feeding a member an `addresses` custom field would pin the chosen behavior.

File: files/netbox/extractors/netplan_extractor.py [hunk 4]
```diff
@@ -672,6 +723,81 @@ def extract(
...
+        # the LAG interface (e.g. to switch to active-backup).
+        for lag_id, lag_interface in lag_interfaces_by_id.items():
+            bond_name = (
+                lag_interface.label if lag_interface.label else lag_interface.name
+            )
+            if not bond_name:
+                continue
+
+            bond_config = {
+                "interfaces": bond_members.get(lag_id, []),
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py]: **Finding 3 (Medium) — Bonds can be emitted with an empty or partial `interfaces` list.**

`bond_members.get(lag_id, [])` can be empty or partial: a member is only collected if it independently passes the tag (`:187`), `mgmt_only` (`:195`), `enabled` (`:428`), and MAC+label (`:416`) gates. The only guard here is `if not bond_name` (`:734`), which doesn't check membership — so a LAG whose members aren't all tagged/labeled emits `interfaces: []`, which is invalid netplan. Please either skip the bond with a warning when membership is empty, or log which members were dropped. Add a test for a managed LAG with no eligible members.

File: files/netbox/extractors/netplan_extractor.py [hunk 5]
```diff
@@ -672,6 +723,81 @@ def extract(
                         f"Added VRF dummy interface {vrf_dummy_name} to VRF {vrf_name} (table {vrf_table}) for device {device.name}"
                     )
 
+        # Process LAG / bond / port channel interfaces now that all members are
+        # known. The default parameters configure an LACP (802.3ad) port channel;
+        # they can be overridden per LAG via the netplan_parameters custom field on
+        # the LAG interface (e.g. to switch to active-backup).
+        for lag_id, lag_interface in lag_interfaces_by_id.items():
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py:736]: **Finding 9 (cleanup, non-blocking) — Boilerplate duplication; natural point to centralize the eligibility gates.**

Non-blocking, but the IP-collection (`:754`), VRF-registration (`:786`), param-merge (`:772`), and MTU-fallback (`:747`) blocks here are each the 5th–7th copy of patterns already in this file. Extracting a few helpers would shrink this and, more usefully, give Findings 1–3 a single shared eligibility gate to fix instead of four call sites. Given the extractor has grown one interface type at a time, this is a reasonable point to pay it down.

File: files/netbox/extractors/netplan_extractor.py [hunk 6]
```diff
@@ -672,6 +723,81 @@ def extract(
...
+                    lag_interface
+                )
+                for ip in ip_addresses:
+                    if ip.address:
+                        addresses.append(ip.address)
+            except Exception:
+                pass
+
+            if addresses:
+                bond_config["addresses"] = addresses
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py:771]: **Finding 6 (OPEN — author design decision) — Bonds never receive the leaf/unnumbered treatment regular ethernets get.**

Regular ethernets connected to a switch with no IPs get the leaf/unnumbered treatment (`link-local: ["ipv6"]`, `dhcp4: False`) at `:472-480`; bonds get nothing equivalent. Is a bonded L3-unnumbered uplink to a leaf/spine an expected topology in this deployment? If yes, this is a latent misconfig and the bond needs the same treatment; if bonds are always IP-carrying here, please confirm so we can note it as intentional.

File: files/netbox/extractors/netplan_extractor.py [hunk 7]
```diff
@@ -672,6 +723,81 @@ def extract(
...
+            # update mirrors the behaviour for other interface types: providing a
+            # "parameters" dict replaces the auto-generated defaults entirely (so
+            # e.g. an active-backup bond does not keep LACP-only options).
+            if (
+                hasattr(lag_interface, "custom_fields")
+                and lag_interface.custom_fields
+            ):
+                lag_params = lag_interface.custom_fields.get("netplan_parameters")
+                if lag_params and isinstance(lag_params, dict):
+                    bond_config.update(lag_params)
```
Comment (ideaship) [inline files/netbox/extractors/netplan_extractor.py]: **Finding 7 (Low — sharp edge) — LAG-level `bond_config.update(lag_params)` can clobber `interfaces`/`mtu`/`addresses`.**

This top-level `update(lag_params)` lets the LAG custom field override not just `parameters` but `interfaces`, `mtu`, and `addresses`. That's consistent with the other interface types and documented as "any other bond key," so not a bug — but a custom-field `interfaces` key would silently replace the auto-detected members. If generated membership is meant to be authoritative, consider excluding `interfaces` from the override; otherwise add a one-line caveat to the docstring/CLAUDE.md noting it can be overridden.

File: files/netbox/CLAUDE.md [hunk 8]
```diff
@@ -117,6 +117,15 @@ The `999-netbox-netplan.yml` file contains netplan_parameters which can be:
     - All IPv4 and IPv6 addresses assigned to the interface are included
     - If the VRF has a table ID, the interface is added to the VRF's interface list in `network_vrfs`
     - The interface is configured in `network_dummy_devices`
+  - **Bond / Port Channel interfaces (LAG)**: Interfaces of type `lag` with the `managed-by-osism` tag (uses the same NetBox modelling as SONiC port channel detection)
+    - The label (or name if no label) becomes the bond name in `network_bonds`
+    - Member interfaces are detected via their `lag` back-reference (i.e. the interface's `lag` field points at the LAG interface)
```
Comment (ideaship) [inline files/netbox/CLAUDE.md]: **Finding 5 (Low — docs) — Member eligibility gates are under-specified (coupled to Findings 1–3).**

This implies the back-reference alone is sufficient, but in practice a member must also independently pass the tag, MAC+label, `enabled`, and `mgmt_only` gates to appear in the bond's `interfaces` — otherwise it's silently dropped (see Findings 1–3). Please document the full member eligibility requirement, and keep this in sync with whatever gating behavior you settle on for the empty/partial-bond and disabled-member cases.

File: tests/unit/netbox/test_netplan_extractor.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+def _extract(interfaces, ips=None):
+    bulk = _FakeBulkLoader(interfaces, ips)
+    client = _FakeNetBoxClient()
+    extractor = NetplanExtractor(api=object(), netbox_client=client, bulk_loader=bulk)
+    return extractor.extract(_device())
+
+
+class TestBondGeneration:
+    def test_lacp_bond_is_generated_from_lag_and_members(self):
```
Comment (ideaship) [inline tests/unit/netbox/test_netplan_extractor.py:96]: **Finding 8 (the de facto blocker) — Test coverage is happy-path only.**

These three tests cover only happy paths. Given the edge cases in this PR, please add tests that pin the intended behavior for: a disabled member of a multi-member bond (Finding 2), a `mgmt_only` and a disabled LAG (Finding 1), a LAG with no eligible members (Finding 3), a member carrying an `addresses`/`match` custom field (Finding 4), and VRF-on-bond vs. VRF-on-member. These tests are what will lock in whatever fix you choose for 1–4.

**Review by ideaship — CHANGES_REQUESTED** [review]

**Review by osfrickler — APPROVED** [review]
IMO the risk of regressions for existing uses is low, so let's merge this and test it on some real deployment
